pyteal/ast/base64decode.py

```diff
@@ -0,0 +1,85 @@
+from typing import TYPE_CHECKING
+from enum import Enum
+
+from pyteal.types import TealType, require_type
+from pyteal.errors import verifyFieldVersion
+from pyteal.ir import TealOp, Op, TealBlock
+from pyteal.ast.expr import Expr
+from pyteal.ast.leafexpr import LeafExpr
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class Base64Encoding(Enum):
+    # fmt: off
+    #     id  |   name   | min version
+    url = (0, "URLEncoding", 7)
+    std = (1, "StdEncoding", 7)
+    # fmt: on
+
+    def __init__(self, id: int, name: str, min_version: int) -> None:
+        self.id = id
+        self.arg_name = name
+        self.min_version = min_version
+
+
+Base64Encoding.__module__ = "pyteal"
+
+
+class Base64Decode(LeafExpr):
+    """An expression that decodes a base64-encoded byte string according to a specific encoding.
+
+    See [RFC 4648](https://rfc-editor.org/rfc/rfc4648.html#section-4) (sections 4 and 5) for information on specifications.
+
+    It is assumed that the encoding ends with the exact number of = padding characters as required by the RFC.
+    When padding occurs, any unused pad bits in the encoding must be set to zero or the decoding will fail.
+    The special cases of \\n and \\r are allowed but completely ignored. An error will result when attempting
+    to decode a string with a character that is not in the encoding alphabet or not one of =, \\r, or \\n.
+    """
+
+    def __init__(self, encoding: Base64Encoding, base64: Expr) -> None:
+        super().__init__()
+        self.encoding = encoding
+
+        require_type(base64, TealType.bytes)
+        self.base64 = base64
+
+    def __teal__(self, options: "CompileOptions"):
+        verifyFieldVersion(
+            self.encoding.arg_name, self.encoding.min_version, options.version
+        )
+
+        op = TealOp(self, Op.base64_decode, self.encoding.arg_name)
+        return TealBlock.FromOp(options, op, self.base64)
+
+    def __str__(self):
+        return "(Base64Decode {})".format(self.encoding.arg_name)
+
+    def type_of(self):
+        return TealType.bytes
+
+    @classmethod
+    def url(cls, base64: Expr) -> "Base64Decode":
```

Comment:
Consider returning `Expr` instead of `Base64Decode`, as that is overly specific.

Same thing for `std`

Reply:
@jasonpaulos I attempted to address in https://github.com/algorand/pyteal/pull/452.  Can you have a look and resolve if the PR captures it?